src/sentry/toolbar/views/iframe_view.py

```diff
@@ -15,77 +16,63 @@ def _get_referrer(request) -> str | None:
+        return super().dispatch(request, *args, **kwargs)
 
-    def handle_not_2fa_compliant(self, request: HttpRequest, *args, **kwargs):
-        return self._handle_logged_out(request, *args, **kwargs)
+    def handle_disabled_member(self, organization: Organization) -> HttpResponse:
+        return self._respond_with_state("logged-out")
 
-    def handle_sudo_required(self, request: HttpRequest, *args, **kwargs):
-        return self._handle_logged_out(request, *args, **kwargs)
+    def handle_not_2fa_compliant(self, request: HttpRequest, *args, **kwargs) -> HttpResponse:
+        return self._respond_with_state("logged-out")
 
-    def handle_auth_required(self, request: HttpRequest, *args, **kwargs):
-        return self._handle_logged_out(request, *args, **kwargs)
+    def handle_sudo_required(self, request: HttpRequest, *args, **kwargs) -> HttpResponse:
+        return self._respond_with_state("logged-out")
 
-    def _handle_logged_out(self, request: HttpRequest, *args, **kwargs):
-        return self.respond(
-            TEMPLATE,
-            status=200,
-            context={
-                "referrer": _get_referrer(request),
-                "state": "logged-out",
-                "logging": request.GET.get("logging", ""),
-            },
-        )
+    def handle_auth_required(self, request: HttpRequest, *args, **kwargs) -> HttpResponse:
+        return self._respond_with_state("logged-out")
 
-    def handle_permission_required(self, request: HttpRequest, *args, **kwargs):
-        return self.respond(
-            TEMPLATE,
-            status=200,
-            context={
-                "referrer": _get_referrer(request),
-                "state": "missing-project",
-                "logging": request.GET.get("logging", ""),
-            },
-        )
+    def handle_permission_required(self, request: HttpRequest, *args, **kwargs) -> HttpResponse:
+        return self._respond_with_state("missing-project")
 
     def get(
         self, request: HttpRequest, organization: Organization, project: Project, *args, **kwargs
-    ):
-        referrer = _get_referrer(request)
+    ) -> HttpResponse:
+        referrer = _get_referrer(request) or ""
         allowed_origins: list[str] = project.get_option("sentry:toolbar_allowed_origins")
 
         if referrer and is_origin_allowed(referrer, allowed_origins):
-            return self.respond(
+            return self._respond_with_state("success")
+
+        return self._respond_with_state("invalid-domain")
+
+    def _respond_with_state(self, state: str):
+        context = {
+            "referrer": _get_referrer(self.request) or "",
+            "state": state,
+            "logging": self.request.GET.get("logging", ""),
+            "organization_slug": self.organization_slug,
+            "project_id_or_slug": self.project_id_or_slug,
+        }
+        return _add_frame_headers(
+            self.request,
+            self.respond(
```

Comment:
```suggestion
            super().respond(
```
Right now it's recursive